qadence/ml_tools/train_grad.py

```
@@ -160,115 +160,178 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
-                        optimizer=optimizer,
-                        loss_fn=loss_fn,
-                        xs=next(dl_iter),  # type: ignore[arg-type]
-                        device=device,
-                        dtype=data_dtype,
-                    )
-
-                else:
-                    raise NotImplementedError(
-                        f"Unsupported dataloader type: {type(dataloader)}. "
-                        "You can use e.g. `qadence.ml_tools.to_dataloader` to build a dataloader."
-                    )
-
-                if (
-                    config.print_every > 0
-                    and iteration % config.print_every == 0
-                    and config.verbose
-                ):
-                    # Note that the loss returned by optimize_step
-                    # is the value before doing the training step
-                    # which is printed accordingly by the previous iteration number
-                    print_metrics(loss, metrics, iteration - 1)
-
-                if config.write_every > 0 and iteration % config.write_every == 0:
-                    write_tracker(
-                        writer, loss, metrics, iteration, tracking_tool=config.tracking_tool
-                    )
-
-                if config.plot_every > 0 and iteration % config.plot_every == 0:
-                    plot_tracker(
-                        writer,
-                        model,
-                        iteration,
-                        config.plotting_functions,
-                        tracking_tool=config.tracking_tool,
-                    )
                 if perform_val:
                     if iteration % config.val_every == 0:
-                        xs = next(dl_iter_val)
-                        xs_to_device = data_to_device(xs, device=device, dtype=data_dtype)
-                        val_loss, *_ = loss_fn(model, xs_to_device)
+                        val_loss, *_ = next_loss_iter(dl_iter_val)
```

we had some bug elsewhere where 0 is a valid value for these `val_every` type arguments, see #511 

IDK if this is guarded already, or if it's worth to add one

---

I think we don't check. But I would raise an error in case it is 0 cause there is no point in defining a disabled callback.